### PR TITLE
Added Api contract test case for users and topologies endpoints.

### PR DESCRIPTION
**Pull request by gokulakrishnansvm**
Added Api contract test case for users and topologies endpoints
<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Ops Test

## What is the best way to verify this PR?

Step 1 : Follow steps from README file(traffic_ops/testing/api_contract) to run api-contract-tests.

Step 2 : To execute servers API Contract Testcase , navigate to traffic_ops/testing/api_contract/v4
pytest --to-user Username --to-password Password --to-url URL --config Config Path --request-template Request template --response-template Response template 

Note: Config, Request template and response template paths are optional. 

## PR submission checklist
- [x] This PR has tests 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 1]
```diff
@@ -311,6 +312,23 @@ def to_login(to_args: ArgsType) -> TOSession:
...
+	Creates new traffic ops db connection.
+	:returns: New Traffic ops database connection
+	"""
+	conn = psycopg2.connect(
+            user="traffic_ops",
+            password="twelve",
+            host="127.0.0.1",
+            port=5432,
+            database="traffic_ops",
+            sslmode="disable"
```
Comment (ericholguin) [inline traffic_ops/testing/api_contract/v4/conftest.py:327]: this DB setup will only work for a local install or an install with that specific username and password, this should be set in our config file instead

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 2]
```diff
@@ -1153,3 +1230,83 @@ def coordinate_data_post(to_session: TOSession, request_template_data: list[JSON
...
+	except KeyError as e:
+		raise TypeError(f"missing user property '{e.args[0]}'") from e
+	user["tenantId"] = tenant_post_data["id"]
+
+	logger.info("New user data to hit POST method %s", user)
+	# Hitting users POST methed
+	response: tuple[JSONData, requests.Response] = to_session.create_user(data=user)
+	resp_obj = check_template_data(response, "user")
+	yield resp_obj
+	coordinate_id = resp_obj.get("id")
```
Comment (ericholguin) [inline traffic_ops/testing/api_contract/v4/conftest.py:1265]: should this be user_id

**Review by ericholguin — CHANGES_REQUESTED** [review]

File: traffic_ops/testing/api_contract/v4/conftest.py (hunk 2 above)
Comment (gokulakrishnansvm) [reply traffic_ops/testing/api_contract/v4/conftest.py:1265]: Yes, It should be user_id, I've changed it.

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 3]
```diff
@@ -1153,3 +1230,83 @@ def coordinate_data_post(to_session: TOSession, request_template_data: list[JSON
...
+	# Create a cursor object to interact with the database
+	cursor = db_connection.cursor()
```
Comment (TerekhovaKate) [inline traffic_ops/testing/api_contract/v4/conftest.py:1267]: explore option around context manager ( with .... operator ) to open and close connection with DB  with the session scope 
`class MyApp(object):
    def __init__(self):
        self.db1 = DB1()
        self.db2 = DB2()
    def __enter__(self):
        self.db1.__enter__()
        try:
            self.db2.__enter__()
        except:
            self.db1.__exit__(None, None, None) # I am not sure with None
            raise
        return self
    def __exit__(self, type, value, traceback):
        try:
            self.db1.__exit__(type, value, traceback)
        finally:
            self.db2.__exit__(type, value, traceback)`

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 4]
```diff
@@ -1153,3 +1230,83 @@ def coordinate_data_post(to_session: TOSession, request_template_data: list[JSON
...
+	cursor.execute("DELETE FROM tm_user WHERE id = %s;", (coordinate_id,))
```
Comment (TerekhovaKate) [inline traffic_ops/testing/api_contract/v4/conftest.py:1268]: do you need `,` after coordinate_id?

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 5]
```diff
@@ -311,6 +312,23 @@ def to_login(to_args: ArgsType) -> TOSession:
...
+	conn = psycopg2.connect(
+            user="traffic_ops",
+            password="twelve",
+            host="127.0.0.1",
+            port=5432,
+            database="traffic_ops",
+            sslmode="disable"
+        )
+	return conn
+
```
Comment (TerekhovaKate) [inline traffic_ops/testing/api_contract/v4/conftest.py:330]: please make sure you have an try except block for DB connection error and error handling 
`try:
    db_connection = psycopg2.connect(**db_info)
    print("Successfully connected to the database.")

except OperationalError:
    print("Error connecting to the database :/")`

**Review by TerekhovaKate — CHANGES_REQUESTED** [review]
Exception Handling Using try and except around DB connection with  psycopg2.

File: traffic_ops/testing/api_contract/v4/conftest.py (hunk 4 above)
Comment (gokulakrishnansvm) [reply traffic_ops/testing/api_contract/v4/conftest.py:1268]: Yes we need that trailing comma, In the context of the execute method,  trailing comma is required when passing a single parameter as a tuple, allows it to be passed as the second argument to the execute method.

File: traffic_ops/testing/api_contract/v4/conftest.py (hunk 1 above)
Comment (gokulakrishnansvm) [reply traffic_ops/testing/api_contract/v4/conftest.py]: Changed it!

File: traffic_ops/testing/api_contract/v4/conftest.py (hunk 5 above)
Comment (gokulakrishnansvm) [reply traffic_ops/testing/api_contract/v4/conftest.py:465]: Added try-except block while creating db connection.

File: traffic_ops/testing/api_contract/v4/conftest.py [hunk 6]
```diff
@@ -313,20 +439,29 @@ def to_login(to_args: ArgsType) -> TOSession:
...
+		to_db_connection = psycopg2.connect(
+            user=to_db_args.user,
+            password=to_db_args.password,
+            host=to_db_args.hostname,
+            port=to_db_args.port,
+            database=to_db_args.db_name,
+            sslmode=to_db_args.sslmode
         )
-	return conn
+		print("Successfully connected to the Traffic Ops database.")
```
Comment (TerekhovaKate) [inline traffic_ops/testing/api_contract/v4/conftest.py:457]: change print to logging

File: traffic_ops/testing/api_contract/v4/conftest.py (hunk 3 above)
Comment (gokulakrishnansvm) [reply traffic_ops/testing/api_contract/v4/conftest.py:1408]: I've explored context manager, but implementation is getting complexed. So, i sticked with current implementation to safely open and close DB connection

**Review by ericholguin — APPROVED** [review]
LGTM

**Pull request merged**